magma/metawfl_handler.py

```diff
@@ -0,0 +1,182 @@
+
+class MetaWorkflowStepSelfDependencyError(ValueError):
+    """Custom ValueError when MetaWorkflow Step has a dependency on itself."""
+    pass
+
+################################################
+#   MetaWorkflowStep
+################################################
+class MetaWorkflowStep(ValidatedDictionary):
+    """
+    Class to represent a MetaWorkflow,
+    as a step within a MetaWorkflow Handler object
+    """
+
+    def __init__(self, input_dict):
+        """
+        Constructor method, initialize object and attributes.
+
+        :param input_dict: a dictionary of MetaWorkflow step metadata
+        :type input_dict: dict
+        """
+        super().__init__(input_dict)
+
+        # Validate presence of basic attributes of this MetaWorkflow step
+        self._validate_basic_attributes(META_WORKFLOW, NAME)
+
+        self._check_self_dependency()
+
+    def _validate_basic_attributes(self, *list_of_attributes):
+        """
+        Validation of the input dictionary for the MetaWorkflow step.
+        Checks that necessary MetaWorkflow attributes are present for this MetaWorkflow step.
+        
+        :param list_of_attributes: attributes that are checked
+        :type list_of_attributes: str(s)
+        :return: None, if all specified attributes are present
+        :raises ValueError: if this object doesn't have a specified attribute
+        :raises AttributeError: if not one (and only one) of items_for_creation attributes is present
+        """
+        super()._validate_basic_attributes(*list_of_attributes)
+        
+        ## Check that one (and only one) of the following attributes is defined on this step:
+        ## ITEMS_FOR_CREATION_UUID or ITEMS_FOR_CREATION_PROP_TRACE
+        try:
+            # set None for [default] arg to not throw AttributeError
+            #TODO: handle this within ff instead? It is CGAP portal-specific
+            if not getattr(self, ITEMS_FOR_CREATION_UUID, None):
+                getattr(self, ITEMS_FOR_CREATION_PROP_TRACE)
+        except AttributeError as e:
+            raise AttributeError("Object validation error, {0}\n"
+                            .format(e.args[0]))
+
+        # for items for creation, this object can only have
+        # either the UUID or property trace, but not both
+        if hasattr(self, ITEMS_FOR_CREATION_PROP_TRACE) and hasattr(self, ITEMS_FOR_CREATION_UUID):
+            raise AttributeError("Object validation error, 'MetaWorkflowStep' object cannot have both of the following attributes: 'items_for_creation_property_trace' and 'items_for_creation_uuid'")
+
+    def _check_self_dependency(self):
+        """
+        Check that this MetaWorkflow Step object doesn't have a self-dependency.
+        
+        :return: None, if no self-dependencies present
+        :raises MetaWorkflowStepSelfDependencyError: if there is a self-dependency
+        """
+        if hasattr(self, DEPENDENCIES):
+            dependencies = getattr(self, DEPENDENCIES)
```

Comment:
Consider condensing to provide a default of an empty list:
```suggestion
        dependencies = getattr(self, DEPENDENCIES, [])
```